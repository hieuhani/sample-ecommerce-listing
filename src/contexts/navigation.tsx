import React, { useReducer, useContext, useMemo } from 'react'
import { ScreenPayload } from '../types'

export interface NavigationProviderProps {
  children: React.ReactNode;
}

export interface State {
  stack: Array<ScreenPayload>;
}

export type Action =
  | { type: 'PUSH'; payload: ScreenPayload }
  | { type: 'POP' };

type Dispatch = (action: Action) => void

const initialState: State = {
  stack: [{
    name: 'PRODUCT_LISTING',
  }]
}

export type ScreenPayloadWithOrder = ScreenPayload & {
  order: number;
}

export type NavigationState = State & {
  currentScreen: ScreenPayload | undefined;
  stackWithOrder: Array<ScreenPayloadWithOrder>;
}

const NavigationStateContext = React.createContext<NavigationState>(initialState as NavigationState)
const NavigationDispatchContext = React.createContext<Dispatch | undefined>(undefined)

export function navigationReducer(state: State, action: Action): State {
  switch (action.type) {
    case 'PUSH':
      return {
       ...state,
       stack: [...state.stack, action.payload],
      }
    case 'POP':
      return {
        ...state,
        stack: state.stack.length > 1 ? state.stack.filter((_, index) => index !== state.stack.length - 1) : initialState.stack
      }
    default:
      throw new Error()
  }
}

export function NavigationProvider({ children }: NavigationProviderProps): React.ReactNode {
  const [state, dispatch] = useReducer(navigationReducer, initialState)
  const navigationState: NavigationState = useMemo(() => {
    return {
      stack: state.stack,
      currentScreen: state.stack.length > 0 ? state.stack.slice(-1)[0] : undefined,
      stackWithOrder: state.stack.map((item, index) => Object.assign({}, item, {
        order: index,
      }))
    }
  }, [state])
  return (
    <NavigationStateContext.Provider value={navigationState}>
      <NavigationDispatchContext.Provider value={dispatch}>
        {children}
      </NavigationDispatchContext.Provider>
    </NavigationStateContext.Provider>
  )
}

export function useNavigationState(): NavigationState {
  const context = useContext(NavigationStateContext)
  if (!context) {
    throw new Error('useNavigationState must be inside a NavigationProvider with a state value')
  }
  return context
}

export function useNavigationDispatch(): Dispatch {
  const context = useContext(NavigationDispatchContext)
  if (!context) {
    throw new Error('useNavigationDispatch must be inside a NavigationProvider with a state dispatch')
  }
  return context
}
