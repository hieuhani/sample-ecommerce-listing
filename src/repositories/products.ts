import { api } from '../services/api'

export interface ProductSearchParams {
  saleCategories?: number;
  channel?: string;
  terminal?: string;
  saleStatuses?: string;
  _sort?: string;
  _order?: 'asc' | 'desc';
  _page?: number;
  _limit?: number;
  publishStatus?: boolean;
  q?: string;
}

export interface ProductPrice {
  supplierSalePrice: number;
  sellPrice: number;
}

export interface ProductPromotionPrice {
  channel: string;
  terminal: string;
  finalPrice: number;
  promotionPrice?: number;
  bestPrice: number;
  flashSalePrice?: number;
}

export interface ProductImage {
  url: string;
  priority: number;
}

export interface Product {
  sku: string;
  displayName: string;
  price: ProductPrice;
  promotionPrices: Array<ProductPromotionPrice>;
  images: Array<ProductImage>;
}

export interface SearchResponseExtra {
  totalItems: number;
  page: number;
  pageSize: number;
}

export interface ProductSearchResponse {
  result: {
    products: Array<Product>;
  };
  extra: SearchResponseExtra;
}

const defaultSearchParams: ProductSearchParams = {
  saleCategories: 613,
  channel: 'pv_online',
  terminal: 'phongvu',
  saleStatuses: 'hang_ban,hang_dat_truoc,hang_sap_het,hang_moi,hang_trung_bay,hang_thanh_ly',
  _sort: 'saleStatuses||hang_ban|hang_dat_truoc|hang_sap_het|hang_moi||hang_trung_bay|hang_thanh_ly||ngung_kinh_doanh',
  _order: 'asc',
  _page: 1,
  _limit: 20,
  publishStatus: true,
}

export function search(params: ProductSearchParams): Promise<ProductSearchResponse> {
  return api.get('/search', Object.assign({}, defaultSearchParams, params)).then(({ ok, data, problem }) => {
    if (!ok) {
      return Promise.reject(problem)
    }
    return Promise.resolve(data as ProductSearchResponse)
  })
}
