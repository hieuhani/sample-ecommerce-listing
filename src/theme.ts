import { DefaultTheme } from 'styled-components'

export const theme: DefaultTheme = {
  colors: {
    reddishOrange: '#f5471e',
    coolGrey: '#8f9598',
    silver: '#c6cace',
    tomato: '#ea341f',
    darkGrey: '#262829',
    paleGrey: '#eef1f3',
    white: '#FFF',
    black30: 'rgba(0, 0, 0, .3)',
    pumpkinOrange: '#f58400',
    deepSkyBlue: '#1e75f6',
  }
}
