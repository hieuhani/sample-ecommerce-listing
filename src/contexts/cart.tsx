import React, { useReducer, useContext, useMemo } from 'react'
import { Product } from '../repositories/products'

export interface CartProviderProps {
  children: React.ReactNode;
}

export interface Cart {
  [key: string]: {
    quantity: number;
    price: number;
  };
}

export interface State {
  cart: Cart;
}

export type Action =
  | { type: 'ADD'; payload: Product }
  | { type: 'DELETE'; payload: Product };

type Dispatch = (action: Action) => void

const initialState: State = {
  cart: {}
}

export type CartState = State & {
  totalItems: number;
  totalValue: number;
}

const CartStateContext = React.createContext<CartState>(initialState as CartState)
const CartDispatchContext = React.createContext<Dispatch | undefined>(undefined)

export function cartReducer(state: State, action: Action): State {
  switch (action.type) {
    case 'ADD': {
      const product = action.payload
      return {
        ...state,
        cart: Object.assign({}, state.cart, {
          [product.sku]: {
            price: product.price.sellPrice,
            quantity: state.cart[product.sku] ? state.cart[product.sku].quantity + 1 : 1,
          }
        })
      }
    }
    case 'DELETE': {
      const product = action.payload
      if (!state.cart[product.sku]) {
        return state
      }
      if (state.cart[product.sku].quantity > 1) {
        return {
          ...state,
          cart: Object.assign({}, state.cart, {
            [product.sku]: {
              price: product.price.sellPrice,
              quantity: state.cart[product.sku].quantity - 1,
            }
          })
        }
      }
      const { [product.sku]: value, ...cart } = state.cart
      return {
        ...state,
        cart,
      }
    }
  }
}

export function CartProvider({ children }: CartProviderProps): React.ReactNode {
  const [state, dispatch] = useReducer(cartReducer, initialState)
  const cartState: CartState = useMemo(() => {
    let totalItems = 0
    let totalValue = 0
    Object.keys(state.cart).forEach(productSku => {
      totalItems += state.cart[productSku].quantity
      totalValue += state.cart[productSku].quantity * state.cart[productSku].price
    })
    return {
      cart: state.cart,
      totalItems,
      totalValue,
    }
  }, [state])
  return (
    <CartStateContext.Provider value={cartState}>
      <CartDispatchContext.Provider value={dispatch}>
        {children}
      </CartDispatchContext.Provider>
    </CartStateContext.Provider>
  )
}

export function useCartState(): CartState {
  const context = useContext(CartStateContext)
  if (!context) {
    throw new Error('useCartState must be inside a CartProvider with a state value')
  }
  return context
}

export function useCartDispatch(): Dispatch {
  const context = useContext(CartDispatchContext)
  if (!context) {
    throw new Error('useCartDispatch must be inside a CartProvider with a state dispatch')
  }
  return context
}
