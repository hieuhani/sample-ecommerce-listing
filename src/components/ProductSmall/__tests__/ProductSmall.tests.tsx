import React from 'react'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { ProductSmall } from '..'
import { theme } from '../../../theme'
import { Product } from '../../../repositories/products'
import { NavigationProvider } from '../../../contexts/navigation'

const product: Product = {
  displayName: 'Máy tính để bàn - PC Acer AS XC-885 (i5-8400/4GB/1TB HDD/UHD 630/Endless)',
  sku: '1807564',
  images: [{
    url: 'https://phongvu.vn/media/catalog_v2/media/48/18/1570775921.343882_Acer_AS_XC-885_1.jpg',
    priority: 1,
  }],
  status: {
    publish: true,
    sale: 'ngung_kinh_doanh' as const,
  },
  promotionPrices: [{
    channel: 'pv_online',
    terminal: 'phongvu',
    finalPrice: 11890000,
    bestPrice: 11890000,
  }],
  price: {
    supplierSalePrice: 11890000,
    sellPrice: 11890000,
  },
  saleCategories: [],
  attributeGroups: [],
  attributes: [],
}

const props = {
  product,
}
describe('ProductSmall', () => {
  test('ProductRow renders correctly', () => {
    const tree = renderer.create(
      <ThemeProvider theme={theme}>
        <NavigationProvider>
          <ProductSmall {...props} />
        </NavigationProvider>
      </ThemeProvider>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
