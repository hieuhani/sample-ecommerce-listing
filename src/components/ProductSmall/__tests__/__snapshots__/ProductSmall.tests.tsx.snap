// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductSmall ProductRow renders correctly 1`] = `
.c5 {
  top: -.3em;
  left: .1em;
}

.c7 {
  color: #8f9598;
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-left: 6pt;
}

.c8 span {
  color: #FFF;
  background: rgb(232,32,32);
  background: linear-gradient(90deg,rgba(232,32,32,1) 0%,#f5471e 100%);
  padding: 2pt 2pt 2pt 0;
  border-top-right-radius: 2pt;
  border-bottom-right-radius: 2pt;
  font-size: 0.8rem;
  margin-left: -6pt;
  z-index: 1001;
}

.c8:before {
  z-index: 1000;
  content: ' ';
  width: 12.3pt;
  height: 12.3pt;
  background: rgb(232,32,32);
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  border-bottom-left-radius: 2pt;
}

.c4 {
  font-weight: 700;
  font-size: 1rem;
  color: #ea341f;
  margin-bottom: 0;
}

.c0 {
  width: 200pt !important;
  padding: 8pt;
}

.c1 {
  height: 200pt;
  width: 200pt;
}

.c1 img {
  width: 100%;
  height: 100%;
}

.c2 {
  font-size: 14pt;
  color: #262829;
  font-weight: 400;
  line-height: 1em;
  height: 2em;
  overflow: hidden;
}

.c6 {
  width: 8pt;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  className="c0 swiper-slide"
  onClick={[Function]}
>
  <div
    className="c1"
  >
    <img
      src="https://phongvu.vn/media/catalog_v2/media/48/18/1570775921.343882_Acer_AS_XC-885_1.jpg"
    />
  </div>
  <h3
    className="c2"
  >
    Máy tính để bàn - PC Acer AS XC-885 (i5-8400/4GB/1TB HDD/UHD 630/Endless)
  </h3>
  <div
    className="c3"
  >
    <h3
      className="c4"
    >
      11.890.000
      <sup
        className="c5"
      >
        đ
      </sup>
    </h3>
    <span
      className="c6"
    />
    <span
      className="c7"
    >
      11.890.000
    </span>
    <span
      className="c6"
    />
    <div
      className="c8"
    >
      <span>
        -
        58
        %
      </span>
    </div>
  </div>
</div>
`;

exports[`ProductSmall ProductRow renders correctly when there are no promotion 1`] = `
.c5 {
  top: -.3em;
  left: .1em;
}

.c4 {
  font-weight: 700;
  font-size: 1rem;
  color: #ea341f;
  margin-bottom: 0;
}

.c0 {
  width: 200pt !important;
  padding: 8pt;
}

.c1 {
  height: 200pt;
  width: 200pt;
}

.c1 img {
  width: 100%;
  height: 100%;
}

.c2 {
  font-size: 14pt;
  color: #262829;
  font-weight: 400;
  line-height: 1em;
  height: 2em;
  overflow: hidden;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
  className="c0 swiper-slide"
  onClick={[Function]}
>
  <div
    className="c1"
  >
    <img
      src="https://phongvu.vn/media/catalog_v2/media/48/18/1570775921.343882_Acer_AS_XC-885_1.jpg"
    />
  </div>
  <h3
    className="c2"
  >
    Máy tính để bàn - PC Acer AS XC-885 (i5-8400/4GB/1TB HDD/UHD 630/Endless)
  </h3>
  <div
    className="c3"
  >
    <h3
      className="c4"
    >
      11.890.000
      <sup
        className="c5"
      >
        đ
      </sup>
    </h3>
  </div>
</div>
`;
